app/Http/Controllers/AlbumController.php

```diff
@@ -541,123 +550,183 @@ public function getArchive(Request $request)
-
-				Logs::notice(__METHOD__, __LINE__, $album->title . ' has been downloaded.');
-
-				//TODO: we do not provide pictures from sub albums but it would be a nice thing to do later...
-
-//				->orWhereIn('album_id',function ($query) { function ($query) use ($id)
-//				{
-//					$query->select('album_id')
-//						->from('user_album')
-//						->where('parent_id','=',$id);
-//				}}));
-				break;
+					break;
+			}
+		} else {
+			$zipTitle = 'Albums';
 		}
 
-		$zipTitle = str_replace($badChars, '', $zipTitle) . '.zip';
-
-		$response = new StreamedResponse(function () use ($zipTitle, $photos_sql, $badChars) {
-			$opt = array(
-				'largeFileSize' => 100 * 1024 * 1024,
-				'enableZip64' => true,
-				'send_headers' => true,
-			);
-
-			$zip = new ZipStream($zipTitle, $opt);
-
-			// Check if album empty
-			if ($photos_sql->count() == 0) {
-				Logs::error(__METHOD__, __LINE__, 'Could not create ZipStream without images');
-
-				return 'false';
-			}
+		$response = new StreamedResponse(function () use ($albumIDs, $badChars) {
+			$zip = new ZipStream(null);
+
+			$UserId = $this->sessionFunctions->id();
+
+			$dirs = [];
+			foreach ($albumIDs as $albumID) {
+				$album = null;
+				switch ($albumID) {
+					case 'f':
+						$dir = 'Starred';
+						if ($this->sessionFunctions->is_logged_in()) {
+							$user = User::find($UserId);
+
+							if ($UserId == 0 || $user->upload) {
+								$photos_sql = Photo::select_stars(Photo::OwnedBy($UserId));
+								break;
+							}
+						}
+						$photos_sql = Photo::select_stars(Photo::whereIn('album_id', $this->albumFunctions->getPublicAlbums()));
+						break;
+					case 's':
+						$dir = 'Public';
+						$photos_sql = Photo::select_public(Photo::OwnedBy($UserId));
+						break;
+					case 'r':
+						$dir = 'Recent';
+						if ($this->sessionFunctions->is_logged_in()) {
+							$user = User::find($UserId);
+
+							if ($UserId == 0 || $user->upload) {
+								$photos_sql = Photo::select_recent(Photo::OwnedBy($UserId));
+								break;
+							}
+						}
+						$photos_sql = Photo::select_recent(Photo::whereIn('album_id', $this->albumFunctions->getPublicAlbums()));
+						break;
+					case '0':
+						$dir = 'Unsorted';
+						$photos_sql = Photo::select_unsorted(Photo::OwnedBy($UserId));
+						break;
+					default:
+						$album = Album::find($albumID);
+						if ($album === null) {
+							Logs::error(__METHOD__, __LINE__, 'Could not find specified album');
+
+							return 'false';
+						}
+						$dir = $album->title;
+						$photos_sql = Photo::set_order(Photo::where('album_id', '=', $albumID));
+						break;
+				} // switch (albumID)
 
-			$photos = $photos_sql->get();
-			foreach ($photos as $photo) {
-				$title = str_replace($badChars, '', $photo->title);
-				$url = Config::get('defines.urls.LYCHEE_URL_UPLOADS_BIG') . $photo->url;
+				$compress_album = function ($photos_sql, $dir, &$dirs, $parent_dir, $album) use (&$zip, $badChars, &$compress_album) {
```

Comment:
Is the use of a functor necessary ? Why not use a private method ?

Reply:
Just laziness I guess :wink:. The code was in the function body so rather than moving it completely out, I created a nested function. I did something similar in the `PhotoController.php` as well -- would you like me to turn them both into private methods?

Reply:
I would be interested to see if those can be delegated to a dependency related to all the archiving stuff like we did for albumfunctions etc... :)